File: QuickUnion/QuickUnionUF.py
class QuickUnionUF:
    def __init__(self, N):
        self.id = list(range(N))

    def root(self, i):
        while i != self.id[i]:
            i = self.id[i]
        return i

    def connected(self, p, q):
        return self.root(p) == self.root(q)

    def union(self, p, q):
        i = self.root(p)
        j = self.root(q)
        self.id[i] = j


uf = QuickUnionUF(10)
uf.union(1, 2)
print(uf.connected(1, 2))  # True
print(uf.connected(1, 3))  # False

File: QuickUnion/QuickUnionUF.js
class QuickUnionUF {
    constructor(N) {
        this.id = Array.from({ length: N }, (_, i) => i);
    }

    root(i) {
        while (i !== this.id[i]) {
            i = this.id[i];
        }
        return i;
    }

    connected(p, q) {
        return this.root(p) === this.root(q);
    }

    union(p, q) {
        const i = this.root(p);
        const j = this.root(q);
        this.id[i] = j;
    }
}


const uf = new QuickUnionUF(10);
uf.union(1, 2);
console.log(uf.connected(1, 2)); // true
console.log(uf.connected(1, 3)); // false
